cmd/gencert/gencert.go

```diff
@@ -0,0 +1,281 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to reconcile certificate secret: %w", err)
+	}
+
+	err = g.reconcileWebhookCA(ctx, certBytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to reconcile webhook ca config: %w", err)
+	}
+
+	return cs[0], nil
+}
+
+func (g *GenCertRunner) getCert(ctx context.Context) ([]byte, []byte, error) {
+	g.Logger.Info("checking for existing secret")
+	var secret corev1.Secret
+	err := g.Client.Get(ctx, types.NamespacedName{Name: g.WebhookTlsSecretName, Namespace: g.Namespace}, &secret)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, nil, err
+	}
+
+	if err == nil {
+		certs, _ := cert.ParseCertsPEM(secret.Data[corev1.TLSCertKey])
+		if len(certs) > 0 && g.needsRenewIn(certs[0], 30*24*time.Hour) > 0 {
+			g.Logger.Info("existing secret does not need to be renewed")
+			return secret.Data[corev1.TLSCertKey], secret.Data[corev1.TLSPrivateKeyKey], nil
+		}
+	}
+
+	g.Logger.Info("creating new self signed secret")
+	return cert.GenerateSelfSignedCertKey(g.dnsNames()[0], nil, g.dnsNames())
+}
+
+func (g *GenCertRunner) reconcileCertSecret(ctx context.Context, certBytes, keyBytes []byte) error {
+	g.Logger.Info("applying cert as secret")
+	secret := &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{Kind: "Secret", APIVersion: corev1.SchemeGroupVersion.String()},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      g.WebhookTlsSecretName,
+			Namespace: g.Namespace,
+		},
+		Type: corev1.SecretTypeTLS,
+		Data: map[string][]byte{
+			corev1.TLSCertKey:       certBytes,
+			corev1.TLSPrivateKeyKey: keyBytes,
+		},
+	}
+
+	return g.Client.Patch(ctx, secret, client.Apply, client.ForceOwnership, client.FieldOwner(vars.FieldOwner))
+}
+
+func (g *GenCertRunner) reconcileWebhookCA(ctx context.Context, caBytes []byte) error {
+	g.Logger.Info("checking webhook configuration")
+	var webhookConfiguration v1.ValidatingWebhookConfiguration
+	err := g.Client.Get(ctx, types.NamespacedName{Name: g.WebhookConfigurationName}, &webhookConfiguration)
+	if err != nil {
+		return err
+	}
+
+	changed := false
+	for i := range webhookConfiguration.Webhooks {
+		whc := &webhookConfiguration.Webhooks[i].ClientConfig
+		if whc.Service == nil {
+			continue
+		}
+
+		if whc.Service.Name != g.WebhookServiceName || whc.Service.Namespace != g.Namespace {
+			continue
+		}
+
+		if string(whc.CABundle) != string(caBytes) {
+			changed = true
+			whc.CABundle = caBytes
+		}
+	}
+
+	if changed {
+		g.Logger.Info("applying changed webhook configuration")
+		return g.Client.Update(ctx, &webhookConfiguration)
+	}
+
+	return nil
+}
+
+func (g *GenCertRunner) needsRenewIn(c *x509.Certificate, renewBefore time.Duration) time.Duration {
+	// There may be repeats in the cert, so we use a set to compare here
+	if !sets.NewString(c.DNSNames...).Equal(sets.NewString(g.dnsNames()...)) {
+		return 0
+	}
+
+	now := time.Now()
+	if now.Before(c.NotBefore) {
+		return 0
+	}
+
+	return c.NotAfter.Sub(now.Add(renewBefore))
```

Comment:
What happens if `now < NotBefore` and `NotAfter < now + renewBefore`? Can only happen if we can have `NotAfter - NotBefore < renewBefore`. That is, a certificate that has a narrow renewal window. Won't this send our reconciler into a busy loop?

Reply:
It should be fine, since the self-signed certs are valid for a year.